File: ruby/structure/sorted_linked_list.rb
class SortedLinkedList
  class Node
    attr_accessor :value, :next_node

    def initialize(value:, next_node: nil)
      @value     = value
      @next_node = next_node
    end
  end

  attr_reader :head, :length

  def initialize(array = nil)
    @head   = nil
    @length = 0

    return if array.nil?
    array.each do |value|
      push(value)
    end
  end

  def first
    first_node&.value
  end

  def last
    last_node&.value
  end

  def at(index)
    node_at(index)&.value
  end

  def push(value)
    node = Node.new(value: value)
    if length == 0
      @head = node
    elsif head.value >= value
      old_head = head
      node.next_node = old_head
      @head = node
    else
      prev_node = head
      prev_node = prev_node.next_node while prev_node.next_node && prev_node.next_node.value < value
      next_node = prev_node.next_node
      prev_node.next_node = node
      node.next_node = next_node
    end
    @length += 1
    value
  end

  def pop
    return nil if length == 0

    node = pre_last_node
    last_value = node.next_node.value
    node.next_node = nil
    @length -= 1
    last_value
  end

  def pop_front
    return nil if length == 0

    old_head = head
    @head = head.next_node
    @length -= 1
    old_head.value
  end

  def remove(position)
    return pop_front if position == 0

    prev_node = node_at(position - 1)
    next_node = prev_node.next_node
    prev_node.next_node = next_node.next_node

    @length -= 1

    next_node.value
  end

  def contains?(value)
    node = head
    while node
      return true if node.value == value
      node = node.next_node
    end
    false
  end

  def to_array
    array = []
    if head
      node = head
      while node
        array << node.value
        node = node.next_node
      end
    end
    array
  end

  private def first_node
    head
  end

  private def last_node
    node = head
    node = node.next_node while node.next_node
    node
  end

  private def pre_last_node
    node = head
    node = node.next_node while node.next_node&.next_node
    node
  end

  private def node_at(index)
    counter = 0
    node = head
    while node && counter < index
      node = node.next_node
      counter += 1
    end
    node
  end
end

# list = SortedLinkedList.new
# fail unless list.to_array == []
# fail unless list.length == 0

# list = SortedLinkedList.new([])
# fail unless list.to_array == []
# fail unless list.length == 0

# list = SortedLinkedList.new([3, 2, 1])
# fail unless list.to_array == [1, 2, 3]
# fail unless list.length == 3
# fail if     list.contains?(0)
# fail unless list.contains?(1)
# fail unless list.contains?(2)
# fail unless list.contains?(3)
# fail if     list.contains?(4)

# list.push(4)
# fail unless list.length == 4
# list.push(0)
# fail unless list.to_array == [0, 1, 2, 3, 4]
# fail unless list.length == 5

# fail unless list.pop == 4
# fail unless list.length == 4
# fail unless list.pop_front == 0
# fail unless list.length == 3

# list.push(9)
# fail unless list.to_array == [1, 2, 3, 9]
# fail unless list.length == 4


# list.remove(0)
# fail unless list.to_array == [2, 3, 9]
# fail unless list.length == 3

# list.remove(1)
# fail unless list.to_array == [2, 9]
# fail unless list.length == 2

# fail unless list.at(0) == 2
# fail unless list.at(1) == 9
# fail unless list.at(2) == nil

File: java/structure/SortedLinkedList.java
import java.util.Iterator;
import java.util.List;
import java.util.ArrayList;

class LinkedList<T extends Comparable<T>> {
  class Node<T> {
    T value;
    Node<T> nextNode;

    Node(T value) {
      this.value = value;
    }
  }

  private class LinkedListIterator<T> implements Iterator<T> {
    private Node<T> current;

    public LinkedListIterator(Node<T> head) {
      current = head;
    }

    @Override
    public void remove() {
      throw new UnsupportedOperationException();
    }

    @Override
    public boolean hasNext() {
      return current != null;
    }

    @Override
    public T next() {
      if (!hasNext()) return null;

      var item = current.value;
      current = current.nextNode;

      return item;
    }
  }

  private Node<T> head;
  int length = 0;

  LinkedList() {
   //
  }

  LinkedList(T[] array) {
    for (int i = 0; i < array.length; ++i) {
      push(array[i]);
    }
  }

  public T first() {
    if (length == 0) return null;
    return head.value;
  }

  public T last() {
    if (length == 0) return null;
    return _lastNode().value;
  }

  public T at(int index) {
    if (length == 0) return null;

    var node = _nodeAt(index);
    if (node == null) return null;
    return node.value;
  }

  public T push(T value) {
    var node = new Node<T>(value);
    if (head == null) {
      head = node;
    } else if (head.value.compareTo(value) >= 0) {
      var oldHead = head;
      node.nextNode = oldHead;
      head = node;
    } else {
      var prevNode = head;
      while (prevNode.nextNode != null && prevNode.nextNode.value.compareTo(value) < 0) {
        prevNode = prevNode.nextNode;
      }
      var nextNode = prevNode.nextNode;
      prevNode.nextNode = node;
      node.nextNode = nextNode;
    }
    length += 1;
    return value;
  }

  public T pop() {
    if (head == null) return null;

    var node = _preLastNode();
    var lastValue = node.nextNode.value;
    node.nextNode = null;
    length -= 1;
    return lastValue;
  }

  public T popFront() {
    if (head == null) return null;

    var oldHead = this.head;
    head = oldHead.nextNode;
    length -= 1;
    return oldHead.value;
  }

  public T remove(int position) {
    if (position == 0) return popFront();

    var prevNode = _nodeAt(position - 1);
    var nextNode = prevNode.nextNode;
    prevNode.nextNode = nextNode.nextNode;
    length -= 1;
    return nextNode.value;
  }

  public Iterator<T> iterator() {
    return new LinkedListIterator<T>(head);
  }

  public Boolean contains(T searchValue) {
    var node = this.head;
    while(node != null) {
      if (node.value == searchValue) return true;
      node = node.nextNode;
    }
    return false;
  }

  public ArrayList<T> toArray() {
    var array = new ArrayList<T>();
    var it = iterator();
    for (int i = 0; i < length; ++i) {
      array.add(it.next());
    }
    return array;
  }

  private Node<T> _lastNode() {
    var node = this.head;
    while(node != null && node.nextNode != null) {
      node = node.nextNode;
    }
    return node;
  }

  private Node<T> _preLastNode() {
    var node = this.head;
    while(node != null && node.nextNode !=null && node.nextNode.nextNode != null) {
      node = node.nextNode;
    }
    return node;
  }

  private Node<T> _nodeAt(int index) {
    int counter = 0;
    var node = this.head;
    while(node != null && counter < index) {
      node = node.nextNode;
      counter += 1;
    }
    return node;
  }

  // public static void main(String[] args) {
  //   var list = new LinkedList<Integer>();
  //   assert list.toArray().equals(List.of());
  //   assert list.length == 0;

  //   Integer[] init = {};
  //   list = new LinkedList<Integer>(init);
  //   assert list.toArray().equals(List.of());
  //   assert list.length == 0;

  //   Integer[] init2 = {3, 2, 1};
  //   list = new LinkedList<Integer>(init2);
  //   assert list.toArray().equals(List.of(1, 2, 3));
  //   assert list.length == 3;
  //   assert !list.contains(0);
  //   assert list.contains(1);
  //   assert list.contains(2);
  //   assert list.contains(3);
  //   assert !list.contains(4);

  //   list.push(4);
  //   assert list.length == 4;
  //   list.push(0);
  //   assert list.toArray().equals(List.of(0, 1, 2, 3, 4));
  //   assert list.length == 5;

  //   assert list.pop() == 4;
  //   assert list.length == 4;
  //   assert list.popFront() == 0;
  //   assert list.length == 3;

  //   list.push(9);
  //   assert list.toArray().equals(List.of(1, 2, 3, 9));
  //   assert list.length == 4;

  //   list.remove(0);
  //   assert list.toArray().equals(List.of(2, 3, 9));
  //   assert list.length == 3;

  //   list.remove(1);
  //   assert list.toArray().equals(List.of(2, 9));
  //   assert list.length == 2;

  //   assert list.at(0) == 2;
  //   assert list.at(1) == 9;
  //   assert list.at(2) == null;
  // }
}
